File: Java/LinkedListCycle.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

public class LinkedListCycle {

    public boolean hasCycle(ListNode head) {
        java.util.HashSet<ListNode> seen = new java.util.HashSet<>();
        ListNode temp = head;
        while (temp != null) {
            if (seen.contains(temp)) {
                return true;
            }
            seen.add(temp);
            temp = temp.next;
        }
        return false;
    }

    public static ListNode buildLinkedList(int[] values, int pos) {
        if (values.length == 0) return null;

        ListNode head = new ListNode(values[0]);
        ListNode curr = head;
        ListNode[] nodes = new ListNode[values.length];
        nodes[0] = head;

        for (int i = 1; i < values.length; i++) {
            nodes[i] = new ListNode(values[i]);
            curr.next = nodes[i];
            curr = curr.next;
        }

        if (pos != -1) {
            curr.next = nodes[pos];
        }

        return head;
    }

    public static void main(String[] args) {
        LinkedListCycle obj = new LinkedListCycle();

        ListNode head1 = buildLinkedList(new int[]{3, 2, 0, -4}, -1);
        System.out.println("Has Cycle (Example 1): " + obj.hasCycle(head1)); // false

        ListNode head2 = buildLinkedList(new int[]{3, 2, 0, -4}, 1);
        System.out.println("Has Cycle (Example 2): " + obj.hasCycle(head2)); // true

    }
}

// Input: head = [3,2,0,-4], pos = 1
// Output: true

// Input: head = [1,2], pos = 0
// Output: true

// Input: head = [1], pos = -1
// Output: false

File: Python/LinkedListCycle.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        temp=head
        dic=set()
        while temp:
            if temp in dic:
                return True
            dic.add(temp)
            temp=temp.next
        return False
    

def buildLinkedList(values, pos=-1):
    if not values:
        return None

    head = ListNode(values[0])
    curr = head
    nodes = [head]

    for val in values[1:]:
        node = ListNode(val)
        curr.next = node
        curr = node
        nodes.append(node)

    if pos != -1:
        curr.next = nodes[pos]

    return head

s=Solution()

head1 = buildLinkedList([3, 2, 0, -4], pos=-1)
print("Has Cycle (Example 1):", s.hasCycle(head1))  # False

head2 = buildLinkedList([3, 2, 0, -4], pos=1)
print("Has Cycle (Example 2):", s.hasCycle(head2))  # True

# Input: head = [3,2,0,-4], pos = 1
# Output: true

# Input: head = [1,2], pos = 0
# Output: true

# Input: head = [1], pos = -1
# Output: false